File: TCPIP/src/22/ping_implementation.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/socket.h>
#include <netinet/ip.h>
#include <netinet/ip_icmp.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <sys/time.h>
#include <errno.h>

#define PACKET_SIZE 64
#define TIMEOUT_SEC 2

// Calculate checksum for ICMP packet
unsigned short checksum(void *b, int len) {
    unsigned short *buf = b;
    unsigned int sum = 0;
    unsigned short result;

    for (sum = 0; len > 1; len -= 2)
        sum += *buf++;
    
    if (len == 1)
        sum += *(unsigned char *)buf;
    
    sum = (sum >> 16) + (sum & 0xFFFF);
    sum += (sum >> 16);
    result = ~sum;
    
    return result;
}

// Send ICMP echo request
int send_ping(int sockfd, struct sockaddr_in *addr, int seq) {
    char packet[PACKET_SIZE];
    struct icmp *icmp_hdr = (struct icmp *)packet;
    
    memset(packet, 0, PACKET_SIZE);
    
    // Fill in ICMP header
    icmp_hdr->icmp_type = ICMP_ECHO;
    icmp_hdr->icmp_code = 0;
    icmp_hdr->icmp_id = getpid();
    icmp_hdr->icmp_seq = seq;
    
    // Fill data section with pattern
    for (int i = sizeof(struct icmp); i < PACKET_SIZE; i++)
        packet[i] = i;
    
    // Calculate checksum
    icmp_hdr->icmp_cksum = 0;
    icmp_hdr->icmp_cksum = checksum(packet, PACKET_SIZE);
    
    // Send packet
    if (sendto(sockfd, packet, PACKET_SIZE, 0, 
               (struct sockaddr *)addr, sizeof(*addr)) <= 0) {
        perror("sendto failed");
        return -1;
    }
    
    return 0;
}

// Receive ICMP echo reply
int receive_ping(int sockfd, struct sockaddr_in *addr, double *rtt) {
    char buffer[1024];
    struct timeval tv_start, tv_end;
    socklen_t addr_len = sizeof(*addr);
    
    gettimeofday(&tv_start, NULL);
    
    int n = recvfrom(sockfd, buffer, sizeof(buffer), 0,
                     (struct sockaddr *)addr, &addr_len);
    
    gettimeofday(&tv_end, NULL);
    
    if (n < 0) {
        if (errno == EAGAIN || errno == EWOULDBLOCK) {
            printf("Request timeout\n");
            return -1;
        }
        perror("recvfrom failed");
        return -1;
    }
    
    // Calculate round-trip time
    *rtt = (tv_end.tv_sec - tv_start.tv_sec) * 1000.0 +
           (tv_end.tv_usec - tv_start.tv_usec) / 1000.0;
    
    // Extract IP and ICMP headers
    struct ip *ip_hdr = (struct ip *)buffer;
    int ip_hdr_len = ip_hdr->ip_hl << 2;
    struct icmp *icmp_hdr = (struct icmp *)(buffer + ip_hdr_len);
    
    // Verify it's an echo reply for our process
    if (icmp_hdr->icmp_type == ICMP_ECHOREPLY && 
        icmp_hdr->icmp_id == getpid()) {
        return icmp_hdr->icmp_seq;
    }
    
    return -1;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <hostname>\n", argv[0]);
        return 1;
    }
    
    // Resolve hostname
    struct hostent *host = gethostbyname(argv[1]);
    if (!host) {
        fprintf(stderr, "Unknown host: %s\n", argv[1]);
        return 1;
    }
    
    struct sockaddr_in addr;
    memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_addr = *(struct in_addr *)host->h_addr;
    
    printf("PING %s (%s): %d data bytes\n", 
           argv[1], inet_ntoa(addr.sin_addr), PACKET_SIZE);
    
    // Create raw socket (requires root/CAP_NET_RAW)
    int sockfd = socket(AF_INET, SOCK_RAW, IPPROTO_ICMP);
    if (sockfd < 0) {
        perror("socket failed (need root privileges)");
        return 1;
    }
    
    // Set socket timeout
    struct timeval timeout;
    timeout.tv_sec = TIMEOUT_SEC;
    timeout.tv_usec = 0;
    setsockopt(sockfd, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));
    
    // Send pings
    int packets_sent = 0, packets_received = 0;
    double total_rtt = 0.0;
    
    for (int seq = 0; seq < 4; seq++) {
        if (send_ping(sockfd, &addr, seq) < 0)
            continue;
        
        packets_sent++;
        
        double rtt;
        int recv_seq = receive_ping(sockfd, &addr, &rtt);
        
        if (recv_seq >= 0) {
            packets_received++;
            total_rtt += rtt;
            printf("%d bytes from %s: icmp_seq=%d ttl=%d time=%.3f ms\n",
                   PACKET_SIZE, inet_ntoa(addr.sin_addr), recv_seq, 64, rtt);
        }
        
        sleep(1);
    }
    
    // Print statistics
    printf("\n--- %s ping statistics ---\n", argv[1]);
    printf("%d packets transmitted, %d received, %.0f%% packet loss\n",
           packets_sent, packets_received, 
           100.0 * (packets_sent - packets_received) / packets_sent);
    
    if (packets_received > 0) {
        printf("rtt min/avg/max = %.3f/%.3f/%.3f ms\n",
               total_rtt / packets_received,
               total_rtt / packets_received,
               total_rtt / packets_received);
    }
    
    close(sockfd);
    return 0;
}

File: TCPIP/src/22/ping_implementation.rs
use std::net::{IpAddr, ToSocketAddrs};
use std::time::{Duration, Instant};
use std::io;

// ICMP Echo Request packet structure
#[repr(C, packed)]
struct IcmpEchoPacket {
    icmp_type: u8,     // Type 8 for echo request
    code: u8,          // Code 0
    checksum: u16,     // Checksum
    identifier: u16,   // Process ID
    sequence: u16,     // Sequence number
    data: [u8; 56],    // Payload data
}

impl IcmpEchoPacket {
    fn new(identifier: u16, sequence: u16) -> Self {
        let mut packet = IcmpEchoPacket {
            icmp_type: 8,
            code: 0,
            checksum: 0,
            identifier: identifier.to_be(),
            sequence: sequence.to_be(),
            data: [0u8; 56],
        };
        
        // Fill data with pattern
        for (i, byte) in packet.data.iter_mut().enumerate() {
            *byte = i as u8;
        }
        
        packet.checksum = packet.calculate_checksum();
        packet
    }
    
    fn calculate_checksum(&self) -> u16 {
        let bytes = unsafe {
            std::slice::from_raw_parts(
                self as *const _ as *const u8,
                std::mem::size_of::<IcmpEchoPacket>(),
            )
        };
        
        let mut sum: u32 = 0;
        let mut i = 0;
        
        // Sum 16-bit words
        while i < bytes.len() - 1 {
            let word = ((bytes[i] as u32) << 8) | (bytes[i + 1] as u32);
            sum += word;
            i += 2;
        }
        
        // Add remaining byte if odd length
        if i < bytes.len() {
            sum += (bytes[i] as u32) << 8;
        }
        
        // Fold 32-bit sum to 16 bits
        while sum >> 16 != 0 {
            sum = (sum & 0xFFFF) + (sum >> 16);
        }
        
        !sum as u16
    }
    
    fn as_bytes(&self) -> &[u8] {
        unsafe {
            std::slice::from_raw_parts(
                self as *const _ as *const u8,
                std::mem::size_of::<IcmpEchoPacket>(),
            )
        }
    }
}

// ICMP Echo Reply parsing
#[repr(C, packed)]
struct IcmpEchoReply {
    icmp_type: u8,
    code: u8,
    checksum: u16,
    identifier: u16,
    sequence: u16,
}

fn send_ping(
    socket: &socket2::Socket,
    addr: &socket2::SockAddr,
    identifier: u16,
    sequence: u16,
) -> io::Result<Instant> {
    let packet = IcmpEchoPacket::new(identifier, sequence);
    let send_time = Instant::now();
    
    socket.send_to(packet.as_bytes(), addr)?;
    
    Ok(send_time)
}

fn receive_ping(
    socket: &socket2::Socket,
    identifier: u16,
    send_time: Instant,
) -> io::Result<(u16, Duration)> {
    let mut buffer = [0u8; 1024];
    
    let (len, _) = socket.recv_from(&mut buffer)?;
    let recv_time = Instant::now();
    
    if len < 28 {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "Packet too short",
        ));
    }
    
    // Skip IP header (typically 20 bytes, but check IHL field)
    let ip_header_len = ((buffer[0] & 0x0F) * 4) as usize;
    
    if len < ip_header_len + 8 {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "ICMP packet too short",
        ));
    }
    
    // Parse ICMP header
    let icmp_data = &buffer[ip_header_len..];
    let icmp_type = icmp_data[0];
    let icmp_id = u16::from_be_bytes([icmp_data[4], icmp_data[5]]);
    let icmp_seq = u16::from_be_bytes([icmp_data[6], icmp_data[7]]);
    
    // Verify it's an echo reply for our request
    if icmp_type == 0 && icmp_id == identifier {
        let rtt = recv_time.duration_since(send_time);
        Ok((icmp_seq, rtt))
    } else {
        Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "Not our echo reply",
        ))
    }
}

fn main() -> io::Result<()> {
    let args: Vec<String> = std::env::args().collect();
    
    if args.len() != 2 {
        eprintln!("Usage: {} <hostname>", args[0]);
        std::process::exit(1);
    }
    
    let host = &args[1];
    
    // Resolve hostname to IP address
    let addr = format!("{}:0", host)
        .to_socket_addrs()?
        .find(|addr| addr.is_ipv4())
        .ok_or_else(|| io::Error::new(io::ErrorKind::NotFound, "Host not found"))?;
    
    println!("PING {} ({}): 64 data bytes", host, addr.ip());
    
    // Create raw ICMP socket (requires CAP_NET_RAW capability)
    let socket = socket2::Socket::new(
        socket2::Domain::IPV4,
        socket2::Type::RAW,
        Some(socket2::Protocol::ICMPV4),
    )?;
    
    // Set receive timeout
    socket.set_read_timeout(Some(Duration::from_secs(2)))?;
    
    let identifier = std::process::id() as u16;
    let sock_addr = socket2::SockAddr::from(addr);
    
    let mut packets_sent = 0;
    let mut packets_received = 0;
    let mut total_rtt = Duration::ZERO;
    
    for seq in 0..4 {
        let send_time = send_ping(&socket, &sock_addr, identifier, seq)?;
        packets_sent += 1;
        
        match receive_ping(&socket, identifier, send_time) {
            Ok((recv_seq, rtt)) => {
                packets_received += 1;
                total_rtt += rtt;
                println!(
                    "64 bytes from {}: icmp_seq={} ttl=64 time={:.3} ms",
                    addr.ip(),
                    recv_seq,
                    rtt.as_secs_f64() * 1000.0
                );
            }
            Err(e) if e.kind() == io::ErrorKind::WouldBlock => {
                println!("Request timeout for icmp_seq {}", seq);
            }
            Err(e) => {
                eprintln!("Error receiving: {}", e);
            }
        }
        
        std::thread::sleep(Duration::from_secs(1));
    }
    
    // Print statistics
    println!("\n--- {} ping statistics ---", host);
    println!(
        "{} packets transmitted, {} received, {:.0}% packet loss",
        packets_sent,
        packets_received,
        100.0 * (packets_sent - packets_received) as f64 / packets_sent as f64
    );
    
    if packets_received > 0 {
        let avg_rtt = total_rtt / packets_received;
        println!(
            "rtt min/avg/max = {:.3}/{:.3}/{:.3} ms",
            avg_rtt.as_secs_f64() * 1000.0,
            avg_rtt.as_secs_f64() * 1000.0,
            avg_rtt.as_secs_f64() * 1000.0
        );
    }
    
    Ok(())
}

// Add to Cargo.toml:
// [dependencies]
// socket2 = "0.5"